mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/configuration/impl/MongoDBConfiguration.java

```diff
@@ -82,14 +86,62 @@ public MongoDBConfiguration(ConfigurationPropertyReader propertyReader, OptionsC
 	 * @return the {@link MongoClientOptions} corresponding to the {@link MongoDBConfiguration}
 	 */
 	public MongoClientOptions buildOptions() {
-		MongoClientOptions.Builder optionsBuilder = new MongoClientOptions.Builder();
+		MongoClientOptions.Builder optionsBuilder = MongoClientOptions.builder();
 
 		optionsBuilder.connectTimeout( timeout );
 		optionsBuilder.writeConcern( writeConcern );
 		optionsBuilder.readPreference( readPreference );
+
+		Map<String, Method> settingsMap = createSettingsMap();
+		for ( Map.Entry<String, Method> entry : settingsMap.entrySet() ) {
+			String setting = MongoDBProperties.MONGO_DRIVER_SETTINGS_PREFIX + "." + entry.getKey();
+			// we know that there is exactly one parameter
+			Class<?> type = entry.getValue().getParameterTypes()[0];
+
+			// for reflection purposes we need to deal with wrapper classes
+			if ( int.class.equals( type ) ) {
+				type = Integer.class;
+			}
+			if ( boolean.class.equals( type ) ) {
+				type = Boolean.class;
+			}
+
+			Object property = propertyReader.property( setting, type ).withDefault( null ).getValue();
+			if ( property == null ) {
+				continue;
+			}
+
+			Method settingMethod = entry.getValue();
+			try {
+				settingMethod.invoke( optionsBuilder, property );
+			}
+			catch ( InvocationTargetException | IllegalAccessException e ) {
+				// TODO OGM-810 fix error message
+			}
+		}
+
 		return optionsBuilder.build();
 	}
 
+	private Map<String, Method> createSettingsMap() {
+		Map<String, Method> settingsMap = new HashMap<>();
+
+		Method[] methods = MongoClientOptions.Builder.class.getDeclaredMethods();
+		for ( Method method : methods ) {
+			if ( method.getParameterTypes().length == 1 ) {
+				Class<?> parameterType = method.getParameterTypes()[0];
+				// we just care of string, int and boolean setters
+				if ( String.class.equals( parameterType )
+						|| int.class.equals( parameterType )
```

Comment:
+1 For removing the old OGM properties for that and expose it via the new settings in the "driver" sub-namespace.

> I think the best solution would be to just have hibernate.ogm.mongodb.driver.writeConcern and handle the cases where the value is a string representation of one of the pre-defined write concern instances, a fully qualified class name or an actual WriteConcern instance.

Yep, that sounds reasonable. `WriteConcern` has several constructors allowing for ad-hoc creation of arbitrarily configured instances, but naturally we cannot support that via text-based configuration properties. So users would have to create specific sub-classes of `WriteConcern` in such cases.

Coding-wise it may be not too bad, check out the concept of `ShortNameProvider` as used in `DatastoreProviderInitiator`. There we also read a setting with known (enum-like) values and a FQN as fallback.

Another thing to consider is the annotation-based config and the programmatic config corresponding to these "write concern"/"read preference" properties (which allow for fine-grained control on the entity/property level). The `@WriteConcern` annotation and the API methods (see `MongoDBGlobalContext` et al.) accept the OGM enum. We meant to not expose the MongoDB-specific types in the API directly, but if we want to support custom write concern settings, I don't think it can be avoided.